Products/GenericSetup/utils.py

```diff
@@ -806,10 +806,11 @@ def _initProperties(self, node):
                     # The type_converters use the ZPublisher default_encoding
                     # for decoding bytes!
                     if self._encoding.lower() != default_encoding:
-                        u_prop_value = prop_value.decode(self._encoding)
-                        prop_value = u_prop_value.encode(default_encoding)
+                        if isinstance(prop_value, six.binary_type):
+                            u_prop_value = prop_value.decode(self._encoding)
+                            prop_value = u_prop_value.encode(default_encoding)
                         prop_value = type_converters[prop_type](prop_value)
-                        if six.PY2:
+                        if six.PY2 and isinstance(prop_value, six.binary_type):
                             u_prop_value = prop_value.decode(default_encoding)
                             prop_value = u_prop_value.encode(self._encoding)
```

Comment:
This looks like madness!  Why are we decoding/re-encoding twice on Python2?

Reply:
Yeah that's true. We are decoding/re-encoding here in general, because `type_converters` expects `utf-8` as `default_encoding` encoded bytes (or something else set in the config). The re-encoding afterwards is necessary to get to the initial encoding of the data. 

But you are probably right with the Python 2 aspect of your comment. If there really is a `binary_type` in Python 3 (which did not occur to me in the tests) there should also be the re-encoding afterwards.

Reply:
``six.binary_type`` is ``str`` in Python 2 and ``bytes`` in Python 3. In other words, in both versions it is a string that has been encoded to some string encoding: https://six.readthedocs.io/#six.binary_type

Reply:
P.S.: Yes, it's madness and I don't want to defend it, but GS is such a complete mess for the Python 3 move that you end up with hacks like these "just to make tests run".  Oftentimes it's not even clear if the test itself is at fault because it tests for something wrong or meaningless.

Reply:
So, do we leave it in the proposed way for now?